text/0160-dispatch-queue.md

```diff
@@ -0,0 +1,66 @@
+- **TEP**: [160](https://github.com/ton-blockchain/TEPs/pull/160) *(don't change)*
+- **title**: Dispatch Queue
+- **status**: Testing
+- **type**: Core
+- **created**: 13.06.2024
+- **replaces**: -
+- **replaced by**: -
+
+# Summary
+From user perspective TON functions as a multithreaded distributed operating system for decentralized applications (dApps). As any operation system TON implements scheduling mechanism for balancing resource consumption across multiple dApps. Previously this balancing was primarly based on sharding, that is moving resource-intensive processes to separate shard, however now we propose to add on top of sharding level balancing, additional introshard balancing, that will prevent a resource monopolization by single dApp that could lead to performance issues for other dApps on the same shard. In particular we propose to add additional intermediate queue between smart-contract and OutMsgQueue - the Dispatch Queue. This intermediate Queue will allow to schedule messages for processing in more distributed manner.
+
+# Motivation
+
+For now the only mechanism of load distribution is sharding: if some shard has too much load it may be divided to two sub-shards. While this indeed helps, there are limitations that sometimes make this mechanims not ideal on practice: finite (and relatively large) time of split/merge of shards, edgecases when load is unevenly distributed inside shard, split limit in network configuration (to make some other node mechanisms predictable) etc. This may affext user experience, in particular predictability and smoothness, since time of chain of transactions execution is dependent on other protocols simultaneously running in the network. The goal of this proposal to better distribute the load and prevent resource monopolization by single, instead providing more predictable execution time for all users and protocols.
+
+# Guide
+
+## Dispatch Queue
+To balance load, we introduce an additional mechanism called the _Dispatch Queue_. When contract sends a message, it may either enter the Dispatch Queue (before the Collator moves some messages from the Dispatch Queue to the OutMsgQueue later on block collation) or directly enter OutMsgQueue. This intermediate step allows for a more even distribution of load among all dApps running in parallel.
+
+The Dispatch Queue is organized as a set of outgoing message queues for each account. When the collator forwards messages to the OutMsgQueue, it maintains lt-order: a message from Account A with `lt1` will move to the OutMsgQueue before a message from Account A with `lt2` if `lt1 < lt2`. However, this order is not maintained for messages from different accounts. For example, a message from Account A with `lt1 > lt2` may be processed before a message from Account B with lt2 if there is a significant backlog of messages from Account B.
```

Comment:
Hello!

The following sentence is not obvious:
>For example, a message from Account A with `lt1 > lt2` may be processed before a message from Account B with lt2 if there is a significant backlog of messages from Account B.

Does it mean, that message came with the earlier `lt` from account `A`, can be processed after, than message with later `lt` came from account `B`? It's strange.

Reply:
Yes, it can, but it is not a new behavior actually.

Imagine situation you have shard (0,2000000000000000) and (0,6000000000000000). Then both shard produce blocks with the same lt_start 48840433000000.
Then lets we have accounts A and B in shard (0,2000000000000000) and account C in shard (0,6000000000000000). And lets account A and C creates two messages to account B with lt 48840433000001 and 48840433000002.

In this case both message A->B will be processed in the same block where it was created, while messages from C will reach account B later (when there will be masterchain block X that reference (0,6000000000000000) and also block in shard (0,2000000000000000) that reference block X).

That leads to the situation when A->B(48840433000002) will be processed earlier than C->B(48840433000001).

The only guarantee we have is that messages of type A->B will be processed in the same order it was sent. This guarantee holds.

Reply:
Thank you for the details and fast reply! Yes, I agree, intershard transactions need more steps comparing to transactions within single shard.

In particular I'm interested in transactions order within a single block and a single shard. Does dispatch queue have impact on transactions order in this case?

More specifically, I have the following transaction chains:
- `1st_external_msg -> A -> B -> Z`
- `2nd_external_msg -> X -> Y -> Z`

I know, that external messages are shuffled, but for example, if `1st_external_msg` is placed earlier after shuffling, will the chain `1st_external_msg -> A -> B -> Z` be processed earlier, than the 2nd chain? Does dispatch queue have impact in this case?

Thank you!